lua/telescope/algos/fzy.lua

```diff
@@ -0,0 +1,180 @@
+    local haystack_chars = {}
+    for i=1,m do
+        haystack_chars[i] = lower_haystack:sub(i, i)
+    end
+
+    for i=1,n do
+        D[i] = {}
+        M[i] = {}
+
+        local prev_score = SCORE_MIN
+        local gap_score = i == n and SCORE_GAP_TRAILING or SCORE_GAP_INNER
+        local needle_char = lower_needle:sub(i, i)
+
+        for j=1,m do
+            if needle_char == haystack_chars[j] then
+                local score = SCORE_MIN
+                if i == 1 then
+                    score = ((j - 1) * SCORE_GAP_LEADING) + match_bonus[j]
+                elseif j > 1 then
+                    local a = M[i - 1][j - 1] + match_bonus[j]
+                    local b = D[i - 1][j - 1] + SCORE_MATCH_CONSECUTIVE
+                    score = math.max(a, b)
+                end
+                D[i][j] = score
+                prev_score = math.max(score, prev_score + gap_score)
+                M[i][j] = prev_score
+            else
+                D[i][j] = SCORE_MIN
+                prev_score = prev_score + gap_score
+                M[i][j] = prev_score
+            end
+        end
+    end
+end
+
+function fzy.score(needle, haystack)
+    local n = string.len(needle)
+    local m = string.len(haystack)
+
+    if n == 0 or m == 0 or m > MATCH_MAX_LENGTH or n > MATCH_MAX_LENGTH then
+        return SCORE_MIN
+    elseif n == m then
+        return SCORE_MAX
+    else
+        local D = {}
+        local M = {}
+        compute(needle, haystack, D, M)
+        return M[n][m]
+    end
+end
+
+function fzy.positions(needle, haystack)
+    local n = string.len(needle)
+    local m = string.len(haystack)
+
+    if n == 0 or m == 0 or m > MATCH_MAX_LENGTH or n > MATCH_MAX_LENGTH then
+        return {}
+    elseif n == m then
+        local consecutive = {}
+        for i=1,n do
+            consecutive[i] = i
+        end
+        return consecutive
+    end
+
+    local D = {}
+    local M = {}
+    compute(needle, haystack, D, M)
+
+    local positions = {}
+    local match_required = false
+    local j = m
+    for i=n,1,-1 do
+        while j >= 1 do
+            if D[i][j] ~= SCORE_MIN and (match_required or D[i][j] == M[i][j]) then
+                match_required = (i ~= 1) and (j ~= 1) and (
+                    M[i][j] == D[i - 1][j - 1] + SCORE_MATCH_CONSECUTIVE)
+                positions[i] = j
+                j = j - 1
+                break
+            else
+                j = j - 1
+            end
+        end
+    end
+
+    return positions
+end
+
+
+fzy.SCORE_GAP_LEADING = SCORE_GAP_LEADING
+fzy.SCORE_GAP_TRAILING = SCORE_GAP_TRAILING
+fzy.SCORE_GAP_INNER = SCORE_GAP_INNER
+fzy.SCORE_MATCH_CONSECUTIVE = SCORE_MATCH_CONSECUTIVE
+fzy.SCORE_MATCH_SLASH = SCORE_MATCH_SLASH
+fzy.SCORE_MATCH_WORD = SCORE_MATCH_WORD
+fzy.SCORE_MATCH_CAPITAL = SCORE_MATCH_CAPITAL
+fzy.SCORE_MATCH_DOT = SCORE_MATCH_DOT
+fzy.SCORE_MAX = SCORE_MAX
+fzy.SCORE_MIN = SCORE_MIN
+fzy.MATCH_MAX_LENGTH = MATCH_MAX_LENGTH
```

Comment:
Should probably not expose these if they aren't actually used:

You need to either use `fzy.CONSTANT` in the actual algorithm or don't expose them, because this doesn't allow people to configure them (which is why I'm assuming you're exposing them).